infer/src/bufferoverrun/absLoc.ml

```diff
@@ -624,21 +615,23 @@ module PowLoc = struct
 
   let cast typ x = map (Loc.cast typ) x
 
-  let to_set = function
-    | Bottom ->
-        LocSet.empty
-    | Unknown ->
-        LocSet.singleton Loc.unknown
-    | Known ploc ->
-        ploc
-
+  let to_set = function Bottom -> LocSet.empty | Unknown ploc | Known ploc -> ploc
 
   let get_linked_list_next ~lhs ~rhs =
     match (is_singleton_or_more lhs, is_singleton_or_more rhs) with
     | Singleton lhs, Singleton rhs ->
         Loc.get_linked_list_next ~lhs ~rhs
     | _, _ ->
         None
+
+
+  let cardinal = function
+    | Bottom ->
+        0
+    | Unknown ploc ->
+        LocSet.cardinal ploc + 1
```

Comment:
This is to ensure that `cardinal` is always greater than 1 for an unknown set. That is, `Unknown {}` is not valid, the unknown set has to always include an unknown location (a location for which `Loc.is_unknown` returns true). This means that we don't force a strong update in https://github.com/facebook/infer/commit/a7b791cee53d70e23aabc34393278ff3b8b419c9.

I see that this is not nice. So I added a new function `PowLoc.is_single_known_loc` and use it instead of checking that `PowLoc.cardinal` is 1.

Reply:
Thank you for the explanation. Now, it is much clearer.